File: tests/refs/kruskal_mst_weight.py
"""Reference: total weight of a minimum spanning tree via Kruskal + union-find."""

from __future__ import annotations
from typing import List, Tuple


class _DSU:
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x: int) -> int:
        root = x
        while self.parent[root] != root:
            root = self.parent[root]
        while self.parent[x] != root:
            nxt = self.parent[x]
            self.parent[x] = root
            x = nxt
        return root

    def union(self, x: int, y: int) -> bool:
        px, py = self.find(x), self.find(y)
        if px == py:
            return False
        if self.rank[px] < self.rank[py]:
            px, py = py, px
        self.parent[py] = px
        if self.rank[px] == self.rank[py]:
            self.rank[px] += 1
        return True


def kruskalMST(V: int, edges: List[Tuple[int, int, int]]) -> int:
    edges_sorted = sorted(edges, key=lambda e: e[2])
    dsu = _DSU(V)
    total = 0
    used = 0
    for u, v, w in edges_sorted:
        if dsu.union(u, v):
            total += w
            used += 1
            if used == V - 1:
                break
    return total

File: tests/refs/KruskalMstWeight.java
/*
 * Reference Java implementation for tests/cases/kruskal_mst_weight.json.
 * Kruskal MST total weight using union-find with path compression.
 */
import java.util.Arrays;
import java.util.Comparator;

public class KruskalMstWeight {
    private static int[] parent;
    private static int[] rank_;

    private static int find(int x) {
        int root = x;
        while (parent[root] != root) root = parent[root];
        while (parent[x] != root) {
            int nxt = parent[x];
            parent[x] = root;
            x = nxt;
        }
        return root;
    }

    private static boolean union(int x, int y) {
        int px = find(x), py = find(y);
        if (px == py) return false;
        if (rank_[px] < rank_[py]) { int t = px; px = py; py = t; }
        parent[py] = px;
        if (rank_[px] == rank_[py]) rank_[px]++;
        return true;
    }

    public static long kruskalMST(long V, long[][] edges) {
        int n = (int) V;
        parent = new int[n];
        rank_ = new int[n];
        for (int i = 0; i < n; ++i) parent[i] = i;
        long[][] sorted = edges.clone();
        Arrays.sort(sorted, Comparator.comparingLong(e -> e[2]));
        long total = 0;
        int used = 0;
        for (long[] e : sorted) {
            if (union((int) e[0], (int) e[1])) {
                total += e[2];
                if (++used == n - 1) break;
            }
        }
        return total;
    }
}
